nexus/src/app/sagas/region_snapshot_replacement_start.rs

```diff
@@ -474,6 +485,94 @@ async fn rsrss_new_region_ensure_undo(
     Ok(())
 }
 
+async fn rsrss_new_region_volume_create(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let osagactx = sagactx.user_data();
+
+    let new_region_volume_id =
+        sagactx.lookup::<Uuid>("new_region_volume_id")?;
+
+    let (new_dataset, ensured_region) = sagactx.lookup::<(
+        db::model::Dataset,
+        crucible_agent_client::types::Region,
+    )>(
+        "ensured_dataset_and_region",
+    )?;
+
+    let Some(new_dataset_address) = new_dataset.address() else {
+        return Err(ActionError::action_failed(format!(
+            "dataset {} does not have an address!",
+            new_dataset.id(),
+        )));
+    };
+
+    let new_region_address = SocketAddrV6::new(
+        *new_dataset_address.ip(),
+        ensured_region.port_number,
+        0,
+        0,
+    );
+
+    // Create a volume to inflate the reference count of the newly created
+    // read-only region. If this is not done it's possible that a user could
+    // delete the snapshot volume _after_ the new read-only region was swapped
+    // in, removing the last reference to it and causing garbage collection.
```

Comment:
Just for my understanding: What prevents the region from being garbage-collected when it's just been ensured but hasn't been linked into anything yet? My guess is that completely orphaned regions don't get GC'ed and that possibly-unused regions are found by looking for them in soft-deleted volumes. Is that correct?

Reply:
Kinda yeah:

- for read-only regions, they're only returned from `soft_delete_volume` for deletion when they appear in the volume being soft-deleted.
- it's similar for read/write regions, but an additional way they're returned for GC is from `find_deleted_volume_regions`. That function join the regions table with the volume table and then only operates on soft-deleted volumes. The relevant section is:

```rust
            // only operate on soft deleted volumes
            let soft_deleted = match &volume {
                Some(volume) => volume.time_deleted.is_some(),
                None => false,
            };

            if !soft_deleted {
                continue;
            }
```

If the region's volume isn't inserted yet, then `volume` will be `None` here.